File: src/main/java/com/leet/medium/MaximalSquare.java
package com.leet.medium;

/**
 * 221.
 *
 * Maximal Square Medium
 *
 * https://leetcode.com/problems/maximal-square/
 *
 * Given an m x n binary matrix filled with 0's and 1's, find the largest square
 * containing only 1's and return its area.
 *
 * Example 1: Input: matrix =
 * [["1","0","1","0","0"],["1","0","1","1","1"],["1","1","1","1","1"],["1","0","0","1","0"]]
 * Output: 4
 *
 * Example 2: Input: matrix = [["0","1"],["1","0"]] Output: 1
 *
 * Example 3: Input: matrix = [["0"]] Output: 0
 *
 * Constraints: m == matrix.length n == matrix[i].length 1 <= m, n <= 300
 * matrix[i][j] is '0' or '1'
 */

public class MaximalSquare {
	public int maximalSquare(char[][] inp) {
		int rows = inp.length, cols = inp[0].length;

		int[] prev = new int[cols + 1];
		int[] curr = new int[cols + 1];

		int ans = 0;

		for (int i = 1; i <= rows; i++) {
			for (int j = 1; j <= cols; j++) {
				if (inp[i - 1][j - 1] == '1') {
					curr[j] = Math.min(prev[j - 1], Math.min(prev[j], curr[j - 1])) + 1;
					ans = Math.max(ans, curr[j]);
				} else {
					curr[j] = 0;
				}
			}

			// swap
			int[] tmp = prev;
			prev = curr;
			curr = tmp;
		}

		return ans * ans;
	}
}

File: src/main/java/com/leet/medium/MaximalSquare.rs
use std::cmp::max;
use std::cmp::min;

pub fn maximal_square(matrix: Vec<Vec<char>>) -> i32 {
  if matrix.is_empty() {
    return 0;
  }

  let rows = matrix.len();
  let cols = matrix[0].len();

  let mut ans = 0;

  // Extra space: O(n)
  let mut running_col: Vec<i32> = vec![0; cols];
  let mut maximals: Vec<i32> = vec![0; cols];

  (0..rows).for_each(|r| {
    let mut running_row = 0;
    let mut diagonal_ans = 0;
    (0..cols).for_each(|c| {
      if matrix[r][c] == '1' {
        let maximal = 1 + min(diagonal_ans, min(running_row, running_col[c]));
        ans = max(maximal, ans);
        running_row += 1;
        running_col[c] += 1;
        diagonal_ans = maximals[c];
        maximals[c] = maximal;
      } else {
        running_row = 0;
        running_col[c] = 0;
        maximals[c] = 0;
      }
    });
  });

  ans * ans
}

pub fn main() {
  let ans = maximal_square(vec![
    vec!['1', '1', '1', '0', '0'],
    vec!['1', '1', '1', '1', '1'],
    vec!['1', '1', '1', '1', '1'],
    vec!['1', '0', '0', '1', '0'],
  ]);

  println!("{}", ans)
}
